src/foam/java/refinements.js

```diff
@@ -1957,57 +1953,54 @@ foam.CLASS({
 
 foam.CLASS({
   package: 'foam.java',
-  name: 'PromisedJavaRefinement',
-  refines: 'foam.core.Promised',
-  flags: [
-    'java'
-  ],
-  methods: [
-    function buildJavaClass(cls) {
-      var name = this.name;
-      var filter = foam.util.flagFilter(['java']);
-      var of = foam.lookup(this.of);
-      var methods = this.methods ?
-        this.methods.map(m => of.getAxiomByName(m)) :
-        of.getOwnAxiomsByClass(foam.core.Method);
-      methods = methods.filter(filter);
-
-      methods.forEach(function(m) {
-        var m2 = m.clone();
-        m2.javaCode = `
+  name: 'PromisedMethodRefinement',
+  refines: 'foam.core.PromisedMethod',
+  flags: ['java'],
+  properties: [
+    {
+      name: 'javaCode',
+      getter: function() {
+        return `
 try {
-  maybeWaitFor${name}();
-  ${m2.javaType != 'void' ? 'return ' : ''}get${foam.String.capitalize(name)}().${m2.name}(${m2.args.map(a => a.name).join(', ')});
+  synchronized ( getObj() ) {
```

Comment:
I'm not sure the value of this synchronized lock?  Shouldn't getObj be synchronized?

Reply:
Oh maybe I'm mistaken but I thought I couldn't do wait() or notifyAll() without being in a synchronized block.